File: Практика/Задание1/ParallelSum.java
package ip.ivanov.danil.Практика.Задание1;

import java.util.concurrent.*;

public class ParallelSum {

    private static final double START = 0;
    private static final double END = 6.28;
    private static final double STEP = 0.0000001;

    public static void main(String[] args) throws InterruptedException, ExecutionException {
        // Создаем пул потоков
        ExecutorService executor = Executors.newFixedThreadPool(2);

        // Сначала вычисляем первую функцию в одном потоке
        double sum1 = calculateSum(START, END, STEP, x -> 34 * x + 0.098);
        System.out.println("Сумма функции 1 (y = 34x + 0.098): " + sum1);

        // Задачи для параллельного вычисления оставшихся функций
        Callable<Double> task2 = () -> calculateSum(START, END, STEP, x -> Math.sin(x) + 45 * Math.pow(x, 2));
        Callable<Double> task3 = () -> calculateSum(START, END, STEP, x -> 1 / (x + 1));

        // Запуск задач
        Future<Double> result2 = executor.submit(task2);
        Future<Double> result3 = executor.submit(task3);

        // Получение результатов
        double sum2 = result2.get();
        double sum3 = result3.get();

        // Закрытие пула потоков
        executor.shutdown();

        // Вычисление среднего значения
        double average = (sum1 + sum2 + sum3) / 3;

        // Вывод результатов
        System.out.println("Сумма функции 2 (y = sin(x) + 45x^2): " + sum2);
        System.out.println("Сумма функции 3 (y = 1 / (x + 1)): " + sum3);
        System.out.println("Среднее значение: " + average);
    }

    // Метод для вычисления суммы значений функции
    private static double calculateSum(double start, double end, double step, java.util.function.Function<Double, Double> function) {
        double sum = 0;
        for (double x = start; x <= end; x += step) {
            sum += function.apply(x);
        }
        return sum;
    }
}

File: Практика/Задание3/parallel_sum.erl
% parallel_sum.erl
        -module(parallel_sum).
        -export([main/0]).

        main() ->
        Start = 0,
        End = 6.28,
        Step = 0.0000001,

        % Запуск первого потока для y = 34x + 0.098
        Pid1 = spawn(fun() -> calculate_sum(Start, End, Step, fun(X -> 34 * X + 0.098 end), self()) end),

        % Получение результата из первого потока
        receive
        {result, Sum1} -> io:format("Сумма функции 1 (y = 34x + 0.098): ~p~n", [Sum1])
        end,

        % Параллельный запуск двух потоков для оставшихся функций
        Pid2 = spawn(fun() -> calculate_sum(Start, End, Step, fun(X -> math:sin(X) + 45 * math:pow(X, 2) end), self()) end),
        Pid3 = spawn(fun() -> calculate_sum(Start, End, Step, fun(X -> 1 / (X + 1) end), self()) end),

        % Получение результатов из двух потоков
        receive
        {result, Sum2} -> io:format("Сумма функции 2 (y = sin(x) + 45x^2): ~p~n", [Sum2])
        end,
        receive
        {result, Sum3} -> io:format("Сумма функции 3 (y = 1 / (x + 1)): ~p~n", [Sum3])
        end,

        % Вычисление среднего значения
        Average = (Sum1 + Sum2 + Sum3) / 3,
        io:format("Среднее значение: ~p~n", [Average]).

        calculate_sum(Start, End, Step, Fun, Parent) ->
        Sum = calculate_sum_loop(Start, End, Step, Fun, 0),
        Parent ! {result, Sum}.

        calculate_sum_loop(X, End, Step, Fun, Acc) when X > End ->
        Acc;
        calculate_sum_loop(X, End, Step, Fun, Acc) ->
        NewAcc = Acc + Fun(X),
        calculate_sum_loop(X + Step, End, Step, Fun, NewAcc).
